core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,167 @@
+
+            if (instanceMetadataWithServicesNotRunning.isEmpty()) {
+                instancesWithServicesNotRunning = Collections.emptyList();
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(instanceMetadataWithServicesNotRunning, stack.getStack());
+                LOGGER.info("Collected {} upscale recovery candidates for hostGroup {}", instancesWithServicesNotRunning.size(),
+                        request.getHostGroupName());
+            }
+
+            StopStartUpscaleGetRecoveryCandidatesResult result =
+                    new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request, instancesWithServicesNotRunning,
+                            request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect upscale recovery candidates for stack: %s, proceeding with subsequent flow.",
+                    request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartUpscaleGetRecoveryCandidatesResult result = new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request,
+                    Collections.emptyList(), request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    private AuthenticatedContext getAuthenticatedContext(StopStartUpscaleGetRecoveryCandidatesRequest request, CloudContext cloudContext,
+            CloudConnector connector) {
+        return connector.authentication().authenticate(cloudContext, request.getCloudCredential());
+    }
+
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses) {
+        Predicate<InstanceMetadataView> instanceHealthCheck = im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            boolean hostUnhealthy = detailedHostStatuses.isHostUnHealthy(hostName);
+            boolean maintenanceMode = detailedHostStatuses.isHostInMaintenanceMode(hostName);
+            boolean hostDecommissioned = detailedHostStatuses.isHostDecommissioned(hostName);
+            boolean servicesUnhealthy = detailedHostStatuses.areServicesUnhealthy(hostName);
+                    // CM: Healthy, Services: Healthy, Host: Unhealthy
```

Comment:
Same comment here as the downscale case.

Reply:
Updated this